### PR TITLE
DataGridLight: Implement master detail

### PR DESCRIPTION
**Pull request by Alyar666**

### CODE REVIEW

File: js/renovation/ui/grids/data_grid_light/master_detail/master_detail.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+  @Template() template!: JSXTemplate<RowTemplateProps>;
+}
+
+@Component({
+  defaultOptionRules: null,
+  view: viewFunction,
+})
+
+export class MasterDetail extends JSXComponent(MasterDetailProps) {
```
Comment (pomahtri) [inline js/renovation/ui/grids/data_grid_light/master_detail/master_detail.tsx]:
```
Warning: MasterDetail component declaration is not correct. Props have required properties. Include their keys to declaration
          MasterDetail extends JSXComponent<MasterDetailProps, "template">
```

Please, fix this warning produced when building components

File: js/renovation/ui/grids/data_grid_light/master_detail/master_detail.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  @Consumer(PluginsContext)
+  plugins = new Plugins();
+
+  @InternalState()
+  keyExpr: KeyExpr = '';
```
Comment (pomahtri) [inline js/renovation/ui/grids/data_grid_light/master_detail/master_detail.tsx]: KeyExpr isn't used in master_detail.tsx

File: js/renovation/ui/grids/data_grid_light/__tests__/test_data.ts [hunk 3]
```diff
@@ -4,4 +4,12 @@ export function generateData(count: number): Record<string, unknown>[] {
   }));
 }
 
+export function generateItems(count: number): any[] {
```
Comment (vconst) [inline js/renovation/ui/grids/data_grid_light/__tests__/test_data.ts]: generateItems -> generateRows

File: js/renovation/ui/grids/data_grid_light/data_grid_light.tsx [hunk 4]
```diff
@@ -94,17 +95,31 @@ export class DataGridLight extends JSXComponent(DataGridLightProps) {
   }
 
   @Effect()
-  updateVisibleItems(): () => void {
-    return this.plugins.watch(VisibleItems, (items) => {
-      this.visibleItems = items;
+  updateVisibleRowsByVisibleItems(): () => void {
+    return this.plugins.watch(VisibleItems, () => {
+      this.visibleRows = this.plugins.getValue(VisibleRows) as Row[];
```
Comment (vconst) [inline js/renovation/ui/grids/data_grid_light/data_grid_light.tsx]: Why  is casting ( as Row[]) necessary here? The VisibleRows value already have type.

File: js/renovation/ui/grids/data_grid_light/data_grid_light.tsx [hunk 5]
```diff
@@ -94,17 +95,31 @@ export class DataGridLight extends JSXComponent(DataGridLightProps) {
...
-    return this.plugins.extend(
-      VisibleItems, -1, () => this.props.dataSource,
-    );
+    return this.plugins.extend(VisibleItems, -1, () => this.props.dataSource);
+  }
+
+  @Effect()
+  setVisibleRowsByVisibleItems(): () => void {
+    return this.plugins.extend(VisibleRows, -1, () => {
+      const visibleItems = this.plugins.getValue(VisibleItems) as RowData[];
```
Comment (vconst) [inline js/renovation/ui/grids/data_grid_light/data_grid_light.tsx]: Why is casting ( as Row[]) necessary here? The VisibleRows value already have type.

File: js/renovation/ui/grids/data_grid_light/data_grid_light.tsx [hunk 6]
```diff
@@ -94,17 +95,31 @@ export class DataGridLight extends JSXComponent(DataGridLightProps) {
...
+
+      return this.processVisibleItems(visibleItems);
```
Comment (vconst) [inline js/renovation/ui/grids/data_grid_light/data_grid_light.tsx]: processVisibleItems -> generateDataRows

File: js/renovation/ui/grids/data_grid_light/data_grid_light.tsx [hunk 7]
```diff
@@ -138,4 +153,14 @@ export class DataGridLight extends JSXComponent(DataGridLightProps) {
       dataField: userColumn,
     }));
   }
+
+  processVisibleItems(visibleItems: RowData[]): Row[] {
+    const keyExpr = this.props.keyExpr as string;
```
Comment (vconst) [inline js/renovation/ui/grids/data_grid_light/data_grid_light.tsx]: Why is casting necessary here?

File: js/renovation/ui/grids/data_grid_light/master_detail/expand_column.tsx [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+  onExpandColumnClick(e: Event): void {
+    const target = e.target as Element;
+
+    if (target.closest('.dx-datagrid-expand')) {
+      this.invertExpanded(this.props.data[this.keyExpr]);
+    }
+  }
+
+  invertExpanded(key: Key): void {
```
Comment (vconst) [inline js/renovation/ui/grids/data_grid_light/master_detail/expand_column.tsx]: invertExpanded -> toggleExpanded

File: js/renovation/ui/grids/data_grid_light/master_detail/expand_column.tsx [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  @Effect()
+  subscribeToRowClick(): () => void {
+    eventsEngine.on(this.cellRef.current, clickEvent, this.onExpandColumnClick);
+    return (): void => eventsEngine.off(this.cellRef.current, clickEvent, this.onExpandColumnClick);
+  }
+
+  onExpandColumnClick(e: Event): void {
+    const target = e.target as Element;
```
Comment (vconst) [inline js/renovation/ui/grids/data_grid_light/master_detail/expand_column.tsx:66]: Is it possible to avoid casting?

**Review by pomahtri — DISMISSED** [review]

**Review by pomahtri — APPROVED** [review]

**Review by vconst — APPROVED** [review]

**Pull request merged**